ksql-common/src/main/java/io/confluent/ksql/util/SchemaUtil.java

```diff
@@ -348,24 +349,57 @@ public static boolean areCompatible(final Schema arg1, final Schema arg2) {
   }
 
   private static boolean mapEquals(final Schema mapA, final Schema mapB) {
-    return Objects.equals(mapA.keySchema(), mapB.keySchema())
-        && Objects.equals(mapA.valueSchema(), mapB.valueSchema());
+    return areCompatible(mapA.keySchema(), mapB.keySchema())
+        && areCompatible(mapA.valueSchema(), mapB.valueSchema());
   }
 
   private static boolean arrayEquals(final Schema arrayA, final Schema arrayB) {
-    return Objects.equals(arrayA.valueSchema(), arrayB.valueSchema());
+    return areCompatible(arrayA.valueSchema(), arrayB.valueSchema());
   }
 
   private static boolean structEquals(final Schema structA, final Schema structB) {
     return structA.fields().isEmpty()
         || structB.fields().isEmpty()
-        || Objects.equals(structA.fields(), structB.fields());
+        || compareFieldsOfStructs(structA, structB);
+  }
+
+
+  private static boolean compareFieldsOfStructs(final Schema structA, final Schema structB) {
+
+    final List<Field> fieldsA = structA.fields();
+    final List<Field> fieldsB = structB.fields();
+    final int sizeA = fieldsA.size();
+    final int sizeB = fieldsB.size();
+
+    if (sizeA != sizeB) {
+      return false;
+    }
+
+    // Custom field comparison to support comparison of structs with decimals and generics
+    for (int i = 0; i < sizeA; i++) {
+      final Field fieldA = fieldsA.get(i);
+      final Field fieldB = fieldsB.get(i);
+      if (!fieldA.name().equals(fieldB.name())
+          || fieldA.index() != fieldB.index()
```

Comment:
I wonder if the order of fields matters 🤔 - it might be better to go through structA and get by name in structB. We can leave that for a follow-up.

Reply:
What I checked is that in Connect, Field equals compares the index of two fields, hence order matters?